hbase-server/src/main/java/org/apache/hadoop/hbase/master/snapshot/SnapshotFileCache.java

```diff
@@ -251,6 +261,31 @@ private void refreshCache() throws IOException {
     this.snapshots.putAll(newSnapshots);
   }
 
+  @VisibleForTesting
+  List<String> getSnapshotsInProgress() throws IOException {
+    List<String> snapshotInProgress = Lists.newArrayList();
+    // only add those files to the cache, but not to the known snapshots
+    Path snapshotTmpDir = new Path(snapshotDir, SnapshotDescriptionUtils.SNAPSHOT_TMP_DIR_NAME);
+    FileStatus[] running = FSUtils.listStatus(fs, snapshotTmpDir);
+    if (running != null) {
+      for (FileStatus run : running) {
+        try {
+          snapshotInProgress.addAll(fileInspector.filesUnderSnapshot(run.getPath()));
+        } catch (CorruptedSnapshotException e) {
+          // See HBASE-16464
+          if (e.getCause() instanceof FileNotFoundException) {
+            // If the snapshot is corrupt, we will delete it
+            fs.delete(run.getPath(), true);
+            LOG.warn("delete the " + run.getPath() + " due to exception:", e.getCause());
```

Comment:
Will this actually work for the ExportSnapshot case? The snapshot manifest is added to tmp before all the files are present on cluster so it looks like this will delete the snapshot manifest which would mess up the import job.

Reply:
Hmmm, there maybe race condition between ExportSnapshot and SnapshotCleaner.
Copying Snapshot Manifest is a fast operation. Maybe we can add a time threshold. When we catch CorruptedSnapshotException, if the modification time of the snapshot folder exceeds a certain time threshold, we will delete it, otherwise we will ignore this cleanup operation. WDYT?

Reply:
copying the snapshot manifest is not always fast since it can be hundreds of MB and the link between clusters can be poor.

Reply:
and when the snapshot contains a large number of files, copying the snapshot can take a long time even when there isn't a lot of data. Also copying the actual data for a large export can take tens-of-days.

Reply:
In fact, when CorruptedSnapshotException is thrown, we can ignore the exception and continue to clean up HFile instead of skip. 

If the CorruptedSnapshotException is thrown, which means that the ExportSnapshot has not copy the snapshot manifest successfully, and the data file of the snapshot has not yet started to copy, so it will have no effect on the snapshot if the snapshotCleaner continues. 

The main purpose of adding a delete snapshot manifest logic is to clean up the abnormal snapshot manifest. Of course, it is OK to not clean it up.

Reply:
Hi @busbey @z-york can you take a look again? Thanks

Reply:
any progress on this issue review?? I faced exactly same issue, and hope it to be resolved.

Reply:
Yeah, if it reads into the middle of copying manifest files, it is ok to remove this snapshot as copying HFiles has not started yet. So there is no impact for the logic in snapshotCleaner.

Reply:
The logic of getUnreferencedFiles() is that for an HFile which is not in cache, it will refreshCache to get the latest snapshot hfiles. If one hfile from this exortSnapshot job is in the list, this means that manifest files have been copied over, so refreshCache() will get the latest snapshot file list.

Reply:
@busbey @z-york Unless you see something missing, I think this one is good to go, thanks.

Reply:
I rebased the patch and posted a new pull request, 
https://github.com/apache/hbase/pull/1791

It is same as the original one, except some minor changes (like some of utilities are moved, change to use new utility class).